add support for meson

stephanlachnit opened the pull request:
I'm currently trying to add a build system for Setzer, so that I can create a .deb package for an official upload in the Debian archive. I think meson is a great choice, it works well with python projects and is rather clean compared to other build systems.

Some notes:
- All the python code currently used should be in the `setzer` namespace for python to prevent conflicts, which is why I moved all python files in the setzer directory (first commit) and then changed how they access the modules (second commit).
- I still need to find a solution for some files in the resources directory.
- ~There is no dependency check (yet).~
- ~I haven't updated the flatpack build.~
- ~I'm not quite sure how the `scripts/generate_symbols.py` file is used.~

I hope you think that adding a build system for distros is a nice idea as well, please let me know.
If you want to try it, just run `meson builddir && ninja -C builddir install`.

cvfosammmm commented:
Thanks a lot for this. Of course I would like to see a Debian package! I can't help with this until Monday though.

generate_symbols.py is not part of the app. It just generates the symbols that are displayed in the sidebar.

stephanlachnit commented:
Cool! I thought a little bit about the resources: right now I just copy the resources folder to the python module folder. This isn't exactly the best style as that directory is intended for code, not data. However, it keeps additional code changes down, so I would leave it as is until you had time to look at the commits (I separated them so you can have a better look at the diffs, would merge them in the end).
A better solution would be to implement a configurable path in `__main__.py`, so that we can change the location of the resources to `share/setzer`. This isn't without disavantages though. Right now, the source tree is still portable, meaning you can run it without any build system or program installation. Using a configurable file means you __need__ to run a build command at least once. This doesn't mean you can't run it without installation, a common method is to build a `setzer.devel` with the local source path as resource path.
At this point the next logical step is to use a gresource file, where one can dump all svgs (or any data files) into one file. It's quite simple to do thanks to meson's gnome module, but it might be useful to add a `setzer.resource` module to simplify the access to the data of that resourc file.
If you have any other ideas, just let me know.

cvfosammmm commented:
I think we can do the resources relocation in a separate pr to keep things simple for now, if this doesn't break anything.

- `generate_symbols.py` should stay in the `scripts` folder.
- The latest commit seems to reverse previous changes (it has the original folder structure). Am I missing something?

stephanlachnit commented:
> I think we can do the resources relocation in a separate pr to keep things simple for now

Sure, did it in #45.

> if this doesn't break anything.

I tried it locally and with an install (note that some Distros (like Debian) only use the `site-packages` python folder in `~/.local`).

> generate_symbols.py should stay in the scripts folder.

Moved it back.

> The latest commit seems to reverse previous changes (it has the original folder structure). Am I missing something?

Which commit do you mean? The flatpak one? Doesn't look like that to me in the diffs but I rebased it anyway on top of master. I'll open the PR, you can use the Squash & Merge functionality of Github to squash them, else I can also squash them if you want.

cvfosammmm commented:
I made a mistake and it's fine. You don't have to squash the commits.

cvfosammmm commented:
`gtk_about_dialog_set_logo ` takes a pixbuf, so this change must be reversed.

stephanlachnit commented:
Ok will look into it. I think there is a way to set just an icon name, which gtk then will take from the standard icon path.

cvfosammmm commented:
Yes there is: `gtk_about_dialog_set_logo_icon_name`.

cvfosammmm commented:
If you make this change and rebase I'm ready to merge.

stephanlachnit commented:
> If you make this change and rebase I'm ready to merge.

Changed it. I ran the About Dialog and the icon appears if I install it, however this does also mean that it doesn't work locally anymore (no crashes, just no icon).
Have you tested the flatpak build? I don't use it so I don't have the build tools, it should work with meson but better safe than sorry.

cvfosammmm commented:
I did a test with flatpak and it works.

I think you should revert the icon change as there doesn't seem to be any benefit, only downside. Ok?

stephanlachnit commented:
> I think you should revert the icon change as there doesn't seem to be any benefit, only downside. Ok?

Then it doesn't work on installation, since the icon location on an install is different. With #45 we could add a workaround that does change the path to include the icon somehow, but I don't see both ways working without a workaround. You can decide what's more important to you, I'll change the PR accordingly.

cvfosammmm commented:
Let's do it in https://github.com/cvfosammmm/Setzer/pull/45 then.
